url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/IntMap.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package edu.columbia.stat.wood.pub.sequencememoizer.util;
+
+import java.io.Serializable;
+
+/**
+ * Array based map for native int keys.
+ * @author nicholasbartlett
+ */
+public class IntMap<E> implements Serializable{
+
+    static final long serialVersionUID = 1;
+
+    private int[] keys;
+    private E[] values;
+
+    /**
+     * Indicates if this map is empty.
+     * @return true if empty, else false
+     */
+    public boolean isEmpty(){
+        return keys == null;
+    }
+
+    /**
+     * Returns the number of mapped keys in the map.
+     * @return size of map
+     */
+    public int size(){
+        if(keys != null){
+            return keys.length;
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * Gets the value associated with the key.
+     * @param key
+     * @return value associated with key, null if no value is found
+     */
+    public E get(int key){
+        if(keys == null || key > keys[keys.length - 1]){
+            return null;
+        } else {
+            int index;
+
+            index = getIndex(key);
+
+            if(keys[index] == key){
+                return values[index];
+            } else {
+                return null;
+            }
+        }
+    }
+
+    /**
+     * Associates a value to a key in the map.
+     * @param key
+     * @param value
+     * @return Object previously assigned to key or null if no value was previously assigned
+     */
+    public E put(int key, E value){
+        if(keys == null){
+            keys = new int[]{key};
+            values = (E[]) new Object[]{value};
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;value&#39; (index: 1 | type: E). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;key&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).